components/sup/src/ctl_gateway/codec.rs

```diff
@@ -0,0 +1,504 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Binary protocol encoder and decoder (codec).
+//!
+//! This module contains functions and types for serializing and deserializing messages sent to
+//! and from sockets speaking `SrvProtocol`. Messages are framed into 3 to 4 segments.
+//!
+//! # Header Segment
+//!
+//! Contains information to a message decoder regarding the count of segments to read and the
+//! length of variable segments. The header segment is a 32 bit integer in big-endian format and
+//! is packed as follows:
+//!
+//! * Header Segment (32-bit)
```

Comment:
Since this protocol is for remote command control, is there any situation in which client and server could be using different versions of the code such that we'd want to version the protocol?

Reply:
You'd want to use the 5 reserved bits in the header. They can be used for any sort of expansion including adding an entire additional segment if you needed it!